### PR TITLE
Make API image smaller by not including dev dependencies

### LINKED ISSUE #1008: Reduce size of API production image
## Problem

Development python packages are included in the final production image for the API <s>and the Ingestion Server</s>. Those aren't needed there and represent a risk; we should be able to remove them easily following a strategy similar to the Catalog with a [docker-compose.yml](https://github.com/WordPress/openverse-catalog/blob/main/docker-compose.yml) / [docker-compose.override.yml](https://github.com/WordPress/openverse-catalog/blob/main/docker-compose.override.yml) files.

## Description

Exclude `dev` packages from the **[openverse-api](https://github.com/WordPress/openverse/pkgs/container/openverse-api)** <s>and [**openverse-ingestion_server**](https://github.com/WordPress/openverse/pkgs/container/openverse-ingestion_server)</s> images.

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes #1008 by @krysal

## Description

This PR separates the API image into `api` and `api_dev`, the former of which is published to GHCR and does not include dev-dependencies, while the latter is used in CI and includes dev-dependencies.

It also updates the `set_matrix_images.py` code to enforce consistency in the outputs so that fallbacks and null-handling is not needed in `ci.yml`.

## Testing Instructions

1. See that CI builds both `api` and `api_dev`.
1. See that CI uses `api_dev` to run the tests (proven by the CI passing, the prod image does not have pytest).
1. See that CI publishes `api` (cannot test without merging the PR unfortunately).

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (`just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`just catalog/generate-docs media-props`
      for the catalog or `just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by dhruvkb** [comment]
It finally works, the tests are passing, and the diff was fairly minimal, but the difference is not as much as I had hoped.

image

I don't know if it's worth proceeding with this.

**Comment by sarayourfriend** [comment]
It's a 7.5% reduction, I'd say that's pretty meaningful.

**Comment by dhruvkb** [comment]
In the CI+CD workflow, we build all the images we would need in `build-images` (this includes `api` and `api_dev` now) and only publish a subset of these images in `publish-images` (this never includes `api-dev`). Images are only built once and then these artifacts are repeatedly downloaded by `load-img` where needed. We build `api_dev` only when the API needs to be tested. For all other scenarios, we only build `api`.

The issue with building a layered image is that we do not include PDM in the final image (just the `.venv/` created in `builder`) so it's not possible to install a PDM lockfile on top of the final image. Additionally installing packages in the `entrypoint` or `cmd` intuitively "feels" wrong.

**Comment by dhruvkb** [comment]
If we build a dev API image in the only place it's used, we can skip the artifact upload, download and cache steps and at the cost of some DRYness, reduce the artifact volume. I'm interested in your thoughts on 2d9dca2.

File: .github/workflows/ci_cd.yml [hunk 1]
```diff
@@ -417,7 +418,34 @@ jobs:
...
+
+      - name: Build API dev image
+        uses: docker/build-push-action@v5
+        with:
+          context: api
+          target: api
+          push: false
+          load: true
+          tags: openverse-api
+          cache-from: type=gha,scope=api_dev
```
Comment (sarayourfriend) [inline .github/workflows/ci_cd.yml:442]: I'm not 100% sure how this works, but if we cache on `scope=api`, would that allow us to reuse the layers from the existing `api` build, if it's there? :thinking: 

A small improvement if it's even possible, so no worries if it isn't or if doing that is very complex.

**Comment by sarayourfriend** [comment]
It looks good to me. Am I correct in understanding that the basic difference is we wouldn't upload/load the `api_dev` image outside that particular API test job, and therefore remove the duplication of an API image loading in every other step?

**Review by sarayourfriend — APPROVED** [review]
I am fine with this increase in complexity. If there are additional improvements we can make towards reusing the layers through shared caching or other approaches, that's fine. Even ignoring the _size_ improvements, it is a basic security improvement not to include dependencies that aren't required for the deployed application. In that sense, this is a win to me regardless of whether it ultimately increases the volume of images/bytes-on-disk/bytes-on-wire.

File: .github/workflows/ci_cd.yml (hunk 1 above)
Comment (dhruvkb) [reply .github/workflows/ci_cd.yml:442]: I was worried that layers from `api-dev` that differ from those in the `api` (especially after the difference of `--dev`/`--prod` in the PDM install step) would overwrite the cache and vice versa making both image builds slower by introducing cache misses.

Anyway, this separate cache automatically speeds up the builds across CI runs so it's still very fast.

**Comment by dhruvkb** [comment]
Yes @sarayourfriend. The `api_dev` image is now only built in the API test job and not cached, uploaded as an artifact or downloaded anywhere.

The production image `api` will be built, uploaded as an artifact, downloaded for use in other jobs and also published to GHCR.

**Review by stacimc — APPROVED** [review]
Interesting discussion in this thread! I agree with the conclusions made here and support the change. Code LGTM. Do you think there's anywhere we should document this change?

**Comment by dhruvkb** [comment]
Yes, there is some documentation about the CI + CD workflow that needs updating. I'll update the docs, push to this PR and merge it.

**Pull request merged**